Kernel/Devices/USB/UHCIController.cpp

```diff
@@ -0,0 +1,470 @@
+
+    m_fullspeed_control_qh->link_next_queue_head(m_bulk_qh);
+    m_fullspeed_control_qh->terminate_element_link_ptr();
+
+    m_bulk_qh->link_next_queue_head(m_dummy_qh);
+    m_bulk_qh->terminate_element_link_ptr();
+
+    auto piix4_td_hack = allocate_transfer_descriptor();
+    piix4_td_hack->terminate();
+    piix4_td_hack->set_max_len(0x7ff); // Null data packet
+    piix4_td_hack->set_device_address(0x7f);
+    piix4_td_hack->set_packet_id(PacketID::IN);
+    m_dummy_qh->terminate_with_stray_descriptor(piix4_td_hack);
+    m_dummy_qh->terminate_element_link_ptr();
+
+    u32* framelist = reinterpret_cast<u32*>(m_framelist->vaddr().as_ptr());
+    for (int frame = 0; frame < UHCI_NUMBER_OF_FRAMES; frame++) {
+        // Each frame pointer points to iso_td % NUM_ISO_TDS
+        framelist[frame] = m_iso_td_list.at(frame % UHCI_NUMBER_OF_ISOCHRONOUS_TDS)->paddr();
+        // klog() << PhysicalAddress(framelist[frame]);
+    }
+
+    m_interrupt_transfer_queue->print();
+    m_lowspeed_control_qh->print();
+    m_fullspeed_control_qh->print();
+    m_bulk_qh->print();
+    m_dummy_qh->print();
+}
+
+QueueHead* UHCIController::allocate_queue_head() const
+{
+    for (QueueHead* queue_head : m_free_qh_pool) {
+        if (!queue_head->in_use()) {
+            queue_head->set_in_use(true);
+#ifdef UHCI_DEBUG
+            klog() << "UHCI: Allocated a new Queue Head! Located @ " << VirtualAddress(queue_head) << "(" << PhysicalAddress(queue_head->paddr()) << ")";
+#endif
+            return queue_head;
+        }
+    }
+
+    ASSERT_NOT_REACHED(); // Let's just assert for now, this should never happen
+    return nullptr;       // Huh!? We're outta queue heads!
+}
+
+TransferDescriptor* UHCIController::allocate_transfer_descriptor() const
+{
+    for (TransferDescriptor* transfer_descriptor : m_free_td_pool) {
+        if (!transfer_descriptor->in_use()) {
+            transfer_descriptor->set_in_use(true);
+#ifdef UHCI_DEBUG
+            klog() << "UHCI: Allocated a new Transfer Descriptor! Located @ " << VirtualAddress(transfer_descriptor) << "(" << PhysicalAddress(transfer_descriptor->paddr()) << ")";
+#endif
+            return transfer_descriptor;
+        }
+    }
+
+    ASSERT_NOT_REACHED(); // Let's just assert for now, this should never happen
+    return nullptr;       // Huh?! We're outta TDs!!
+}
+
+void UHCIController::stop()
+{
+    write_usbcmd(read_usbcmd() & ~UHCI_USBCMD_RUN);
+    // FIXME: Timeout
+    for (;;) {
+        if (read_usbsts() & UHCI_USBSTS_HOST_CONTROLLER_HALTED)
+            break;
+    }
+}
+
+void UHCIController::start()
+{
+    write_usbcmd(read_usbcmd() | UHCI_USBCMD_RUN);
+    // FIXME: Timeout
+    for (;;) {
+        if (!(read_usbsts() & UHCI_USBSTS_HOST_CONTROLLER_HALTED))
+            break;
+    }
+    klog() << "UHCI: Started!";
+}
+
+struct setup_packet {
+    u8 bmRequestType;
+    u8 bRequest;
+    u16 wValue;
+    u16 wIndex;
+    u16 wLength;
+};
+
+void UHCIController::do_debug_transfer()
+{
```

Comment:
Stuff like this will be removed next PR. More of a baseline for other people to see how this is done if they want to join in on working on USB stuff, because the spec does a very poor job at actually explaining how these are performed at all.